pkg/services/ngalert/state/state.go

```diff
@@ -129,83 +180,77 @@ func NewEvaluationValues(m map[string]eval.NumberValueCapture) map[string]*float
-			state.State = eval.Alerting
-			state.StartsAt = result.EvaluatedAt
-			state.setEndsAt(rule, result)
+			state.Alerting(result.EvaluatedAt, nextEndsTime(rule.IntervalSeconds, result.EvaluatedAt))
 		}
 	default:
 		if rule.For > 0 {
 			// If the alert rule has a For duration that should be observed then the state should be set to Pending
 			logger.Debug("Changing state", "previous_state", state.State, "next_state", eval.Pending)
-			state.State = eval.Pending
+			state.Pending(result.EvaluatedAt, nextEndsTime(rule.IntervalSeconds, result.EvaluatedAt))
 		} else {
 			logger.Debug("Changing state", "previous_state", state.State, "next_state", eval.Alerting)
-			state.State = eval.Alerting
+			state.Alerting(result.EvaluatedAt, nextEndsTime(rule.IntervalSeconds, result.EvaluatedAt))
 		}
-		state.StartsAt = result.EvaluatedAt
-		state.setEndsAt(rule, result)
 	}
 }
-
 func resultError(state *State, rule *models.AlertRule, result eval.Result, logger log.Logger) {
 	switch rule.ExecErrState {
 	case models.AlertingErrState:
+		logger.Debug("Execution error state is Alerting", "handler", "resultAlerting", "previous_handler", "resultError")
 		resultAlerting(state, rule, result, logger)
-	case models.ErrorErrState:
+		// This is a special case where Alerting and Pending should also have an error and reason
 		state.Error = result.Error
-		if result.Error != nil {
-			// If the evaluation failed because a query returned an error then add the Ref ID and
-			// Datasource UID as labels
-			var queryError expr.QueryError
-			if errors.As(state.Error, &queryError) {
-				for _, next := range rule.Data {
-					if next.RefID == queryError.RefID {
-						state.Labels["ref_id"] = next.RefID
-						state.Labels["datasource_uid"] = next.DatasourceUID
-						break
-					}
-				}
-				state.Annotations["Error"] = queryError.Error()
-			}
-		}
-
+		state.StateReason = "error"
+	case models.ErrorErrState:
 		if state.State == eval.Error {
-			// If the previous state is Error then update the expiration time
-			state.setEndsAt(rule, result)
+			logger.Debug("Keeping state", "state", state.State)
+			state.Maintain(rule.IntervalSeconds, result.EvaluatedAt)
 		} else {
 			// This is the first occurrence of an error
 			logger.Debug("Changing state", "previous_state", state.State, "next_state", eval.Error)
-			state.State = eval.Error
-			state.StartsAt = result.EvaluatedAt
-			state.setEndsAt(rule, result)
+			state.SetError(result.Error, result.EvaluatedAt, nextEndsTime(rule.IntervalSeconds, result.EvaluatedAt))
+
+			if result.Error != nil {
+				// If the evaluation failed because a query returned an error then add the Ref ID and
+				// Datasource UID as labels
+				var queryError expr.QueryError
+				if errors.As(state.Error, &queryError) {
+					for _, next := range rule.Data {
+						if next.RefID == queryError.RefID {
+							state.Labels["ref_id"] = next.RefID
+							state.Labels["datasource_uid"] = next.DatasourceUID
```

Comment:
This is actually a bug because we add labels without proper calculation of the key (which is a result of joining labels).
If we want to create a state per datasource_uid, we need to demux the error to multiple results somewhere before Manager.setNextState and the go the regular way.

Reply:
Also, this logic resembles the one we do for NoData, which lives in eval package. I think this logic should live there as well.

Reply:
We discussed this offline. The idea is that we should not mutate labels because they define the unique key of the state. An alternative could be an annotation, though. However, I think it makes sense to create a state per data source UID.